pkg/network/notarypool/notary_pool.go

```diff
@@ -0,0 +1,246 @@
+	lock        sync.RWMutex
+	verifiedMap map[util.Uint256]item
+	fees        map[util.Uint160]utilityBalanceAndFees
+	capacity    int
+
+	resendThreshold uint32
+	resendFunc      func(*payload.P2PNotaryRequest)
+
+	notaryContractHash util.Uint160
+	policer            blockchainer.Policer
+}
+
+var (
+	// ErrInsufficientDeposit is returned when sender does not have enough GAS deposited to the
+	// Notary contract account to pay all the fallback transactions' fees.
+	ErrInsufficientDeposit = errors.New("insufficient deposit amount")
+	// ErrOOM is returned when NotaryPool capacity is exceeded.
+	ErrOOM = errors.New("out of memory")
+	// ErrAlreadyExists is returned when payload is already in the pool.
+	ErrAlreadyExists = errors.New("payload already in pool")
+)
+
+// NewNotaryPool returns a new NotaryPool struct.
+func NewNotaryPool(capacity int, bc blockchainer.Blockchainer) *NotaryPool {
+	notaryContractHash, _ := bc.GetNativeContractScriptHash("notary") // we are sure that it is exists.
+	return &NotaryPool{
+		verifiedMap:        make(map[util.Uint256]item),
+		fees:               make(map[util.Uint160]utilityBalanceAndFees),
+		capacity:           capacity,
+		notaryContractHash: notaryContractHash,
+		policer:            bc.GetPolicer(),
+	}
+}
+
+// HasPayload checks whether payload with the specified hash is in the pool.
+func (mp *NotaryPool) HasPayload(h util.Uint256) bool {
+	mp.lock.RLock()
+	defer mp.lock.RUnlock()
+	return mp.containsKey(h)
+}
+
+// GetPayload returns payload with the specified hash if it is in the mempool.
+func (mp *NotaryPool) GetPayload(h util.Uint256) *payload.P2PNotaryRequest {
+	mp.lock.RLock()
+	defer mp.lock.RUnlock()
+	return mp.verifiedMap[h].payload
+}
+
+// VerifyAndPoolPayload is a callback method, need to be called under the blockchain lock.
+func (mp *NotaryPool) VerifyAndPoolPayload(bc blockchainer.Blockchainer, r *payload.P2PNotaryRequest) error {
+	payer := r.FallbackTransaction.Signers[1].Account
+	if err := bc.VerifyWitness(payer, r, &r.Witness, mp.policer.GetMaxVerificationGAS()); err != nil {
+		return fmt.Errorf("bad P2PNotaryRequest payload witness: %w", err)
+	}
+	if r.FallbackTransaction.Sender() != mp.notaryContractHash {
+		return errors.New("P2PNotary contract should be a sender of the fallback transaction")
+	}
+	if err := bc.VerifyWitness(payer, r.FallbackTransaction, &r.FallbackTransaction.Scripts[1], mp.policer.GetMaxVerificationGAS()); err != nil {
+		return fmt.Errorf("bad fallback transaction witness #1: %s", err.Error())
+	}
+	nvbFallback := r.FallbackTransaction.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height
+	maxNVBDelta := mp.policer.GetMaxNotValidBeforeDelta()
+	if bc.BlockHeight()+maxNVBDelta < nvbFallback {
+		return fmt.Errorf("fallback transaction should have NotValidBefore not more then %d, got %d", bc.BlockHeight()+maxNVBDelta, nvbFallback)
+	}
+	if nvbFallback+maxNVBDelta < r.FallbackTransaction.ValidUntilBlock { // TODO: <= ?
```

Comment:
There's one last question, don't we want to make here `<=`?

Reply:
For chain height H the transaction is valid if `NVB == H` and not valid if `VUB == H`. Here we want transaction to be valid for `maxNVBDelta` number of blocks, I think we're getting this property with `<`, so it should be fine.